test/bio_dgram_test.c

```diff
@@ -294,6 +295,282 @@ static int test_bio_dgram(void)
 
 #endif
 
+static int random_data(const uint32_t *key, uint8_t *data, size_t data_len, size_t offset)
+{
+    int ret = 0, outl;
+    EVP_CIPHER_CTX *ctx = NULL;
+    EVP_CIPHER *cipher = NULL;
+    static const uint8_t zeroes[2048];
+    uint32_t counter[4] = { (uint32_t)offset };
+
+    ctx = EVP_CIPHER_CTX_new();
+    if (ctx == NULL)
+        goto err;
+
+    cipher = EVP_CIPHER_fetch(NULL, "ChaCha20", NULL);
+    if (cipher == NULL)
+        goto err;
+
+    if (EVP_EncryptInit_ex2(ctx, cipher, (uint8_t *)key, (uint8_t *)counter, NULL) == 0)
+        goto err;
+
+    while (data_len > 0) {
+        outl = data_len > sizeof(zeroes) ? (int)sizeof(zeroes) : (int)data_len;
+        if (EVP_EncryptUpdate(ctx, data, &outl, zeroes, outl) != 1)
+            goto err;
+
+        data     += outl;
+        data_len -= outl;
+    }
+
+    ret = 1;
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(cipher);
+    return ret;
+}
+
+static int test_bio_dgram_pair(void)
+{
+    int testresult = 0, blen, mtu1, mtu2, r;
+    BIO *bio1 = NULL, *bio2 = NULL;
+    uint8_t scratch[2048+4], scratch2[2048];
+    uint32_t key[8];
+    size_t i, num_dgram;
+    BIO_MSG msgs[2] = {0}, rmsgs[2] = {0};
+    BIO_ADDR *addr1 = NULL, *addr2 = NULL, *addr3 = NULL, *addr4 = NULL;
+    struct in_addr in_local = {ntohl(0x7f000001)};
+    const uint32_t ref_caps = BIO_DGRAM_CAP_HANDLES_SRC_ADDR
+                            | BIO_DGRAM_CAP_HANDLES_DST_ADDR
+                            | BIO_DGRAM_CAP_PROVIDES_SRC_ADDR
+                            | BIO_DGRAM_CAP_PROVIDES_DST_ADDR;
+
+    if (!TEST_int_eq(RAND_bytes((uint8_t *)key, sizeof(key)), 1))
+        goto err;
```

Comment:
This test is problematic because it is not repeatable. If a bug is data dependent in some way then this test could fail sometimes and pass on other times. Can we make it repeatable somehow?